File: 17-12/Inorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root):  # 66ms
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        self.helper(root,res)
        return res

    def helper(self,node,res):
        if node:
            self.helper(node.left,res)
            res.append(node.val)
            self.helper(node.right,res)

    def iterative_inorder_traversal(self,root):  # 59ms
        res, stack = [], []
        while True:
            while root:
                stack.append(root)
                root = root.left
            if not stack:
                return res
            node = stack.pop()
            res.append(node.val)
            root = node.right

File: 19-09/treetraverse/inorder/inorder.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recursive(root *TreeNode) []int {
	result := []int{}
	helper(root, &result)
	return result
}

func helper(node *TreeNode, order *[]int) {
	if node == nil {
		return
	}
	helper(node.Left, order)
	*order = append(*order, node.Val)
	helper(node.Right, order)
}

func inorderTraversal(root *TreeNode) []int {
	var result []int
	cur := root
	var stack []*TreeNode
	for cur != nil || len(stack) > 0 {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		cur = stack[len(stack)-1]
		result = append(result, cur.Val)
		stack = stack[:len(stack)-1]
		cur = cur.Right
	}
	return result
}
